### PR TITLE
Attach policy labels to Messages

### PR DESCRIPTION
**Pull request by tiziano88**
Currently those labels are not used for anything, but in the future they
will be used for information flow.

Convert constant from absl::string_view back to char*, it does not seem
to work with std::multimap otherwise.

Rename _policy_ to _label_. This may still not be the final terminology,
but seems more appropriate.

Reintroduce `--output_base` flag to `build_example` to avoid losing
incremental state.

Ref #212

### CODE REVIEW

**Comment by tiziano88** [comment]
cc @aferr

File: oak/server/module_invocation.cc [hunk 1]
```diff
@@ -77,11 +81,19 @@ void ModuleInvocation::ProcessRequest(bool ok) {
...
+  std::unique_ptr<Message> req_msg = absl::make_unique<Message>();
+  req_msg->data.insert(req_msg->data.end(), encap_req.begin(), encap_req.end());
+  {
+    auto range = context_.client_metadata().equal_range(kOakLabelMetadataKey);
+    for (auto entry = range.first; entry != range.second; ++entry) {
+      std::string label(entry->second.data(), entry->second.size());
+      LOG(INFO) << "Oak label: " << label;
+      req_msg->labels.push_back(label);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/module_invocation.cc]: I've not looked at this yet, but we should sync up &ndash; I've been experimenting with propagating labels between Nodes and messages in the TCB: https://github.com/daviddrysdale/oak/commits/label especially [4ffecb3f2a7](https://github.com/daviddrysdale/oak/commit/4ffecb3f2a703e67e5f916068cbf60b87d71eb85)

File: oak/server/channel.h [hunk 2]
```diff
@@ -39,9 +39,11 @@ using ChannelHalf = absl::variant<std::unique_ptr<MessageChannelReadHalf>,
 
 // Each message transferred over a channel includes data and potentially
 // also references to halves of channels.
+// It also has a set of labels attached to it, which are used for information flow.
 struct Message {
   std::vector<char> data;
   std::vector<std::unique_ptr<ChannelHalf>> channels;
+  std::vector<std::string> labels;
```
Comment (daviddrysdale) [inline oak/server/channel.h]: (I've got something similar to this in #273)

File: oak/server/module_invocation.cc [hunk 3]
```diff
@@ -77,11 +81,19 @@ void ModuleInvocation::ProcessRequest(bool ok) {
   std::string encap_req;
   grpc_request.SerializeToString(&encap_req);
   // TODO: figure out a way to avoid the extra copy (into then out of std::string)
-  std::unique_ptr<Message> rsp_msg = absl::make_unique<Message>();
-  rsp_msg->data.insert(rsp_msg->data.end(), encap_req.begin(), encap_req.end());
+  std::unique_ptr<Message> req_msg = absl::make_unique<Message>();
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/module_invocation.cc]: (personal preference) Maybe fix it in a distinct commit?

File: oak/server/module_invocation.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/module_invocation.cc]: My experiment is now in #273, should be straightforward to merge with this once we figure the right mapping between 'policy' and 'secrecy'/'integrity'.

File: oak/server/BUILD [hunk 4]
```diff
@@ -115,6 +115,7 @@ cc_library(
     deps = [
         ":channel",
         ":oak_node",
+        "//oak/client:policy_metadata",
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/server/BUILD]: Done.

File: oak/server/module_invocation.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/server/module_invocation.cc]: Done.

File: scripts/build_example [hunk 5]
```diff
@@ -50,7 +50,9 @@ bazel_build_flags+=(
     "--compilation_mode=$COMPILATION_MODE"
 )
 
-bazel build "${bazel_build_flags[@]}" "//examples/$TARGET/client"
+# Use a different output_base so that we don't lose incremental state.
+# See https://docs.bazel.build/versions/master/command-line-reference.html#flag--output_base.
+bazel --output_base="$CACHE_DIR/client" build "${bazel_build_flags[@]}" "//examples/$TARGET/client"
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply scripts/build_example]: Done.

File: oak/client/BUILD [hunk 6]
```diff
@@ -50,6 +50,7 @@ cc_library(
     hdrs = ["policy_metadata.h"],
     visibility = ["//visibility:public"],
     deps = [
+        "//oak/common:policy",
         "@com_github_grpc_grpc//:grpc++",
         "@com_google_absl//absl/base",
         "@com_google_absl//absl/strings",
```
Comment (daviddrysdale) [inline oak/client/BUILD]: Is the `absl/strings` dep still needed?

File: oak/server/module_invocation.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/module_invocation.cc]: nit: rename of `rsp_msg` -> `req_msg` still commingled with policy changes.

File: oak/server/module_invocation.cc [hunk 7]
```diff
@@ -77,11 +81,19 @@ void ModuleInvocation::ProcessRequest(bool ok) {
...
+  req_msg->data.insert(req_msg->data.end(), encap_req.begin(), encap_req.end());
+  {
```
Comment (daviddrysdale) [inline oak/server/module_invocation.cc]: Why the block?

Comment (tiziano88) [reply oak/server/module_invocation.cc]: I like to use blocks to minimize scope of auxiliary variables, such as `range` in this context.

File: oak/server/channel.h (hunk 2 above)
Comment (tiziano88) [reply oak/server/channel.h]: Feel free to replace it entirely when you merge your changes, this is definitely just a placeholder.

File: oak/client/BUILD (hunk 6 above)
Comment (tiziano88) [reply oak/client/BUILD]: Removed.

**Pull request merged**